%\VignetteIndexEntry{Genomic Annotation Practical}
%\VignettePackage{GeneticsHTSCourse}
%\VignetteEngine{knitr::knitr}

% To compile this document
% library('knitr'); rm(list=ls()); knit('DESeq2.Rnw')

\documentclass[12pt]{article}
\newcommand{\usecase}{\textit{\textbf{Use Case: }}}

<<knitr, echo=FALSE, results="hide">>=
library("knitr")
opts_chunk$set(tidy=FALSE,dev="png",fig.show="hide",
               fig.width=4,fig.height=4.5,
               message=FALSE,eval=F)
@ 

<<style, eval=TRUE, echo=FALSE, results="asis">>=
BiocStyle::latex()
@


\title{Annotation and Visualisation of sequencing data in Bioconductor}

\author{Mark Dunning}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{document}

\maketitle
\tableofcontents
\section{Introduction}

This practical will use the data from the Pasilla dataset, which you will have already analysed in the RNA-seq pratical. We will assume that the gene counts data have already been analysed. \textbf{You do not need to type this code, it is for your reference only.}

<<>>=
library(DESeq)

datafile = system.file( "extdata/pasilla_gene_counts.tsv", package="pasilla" )
pasillaCountTable = read.table( datafile, header=TRUE, row.names=1 )
pasillaDesign = data.frame(row.names = colnames( pasillaCountTable ),
condition = c("untreated","untreated","untreated","untreated",
  "treated","treated","treated"),
libType=c("single-end","single-end", "paired-end","paired-end","single-end", "paired-end","paired-end"))
pairedSamples = pasillaDesign$libType == "paired-end"
condition = factor( c( "untreated", "untreated", "treated", "treated" ) )
countTable = pasillaCountTable[ , pairedSamples ]
condition = factor( c( "untreated", "untreated", "treated", "treated" ) )
cds = newCountDataSet( countTable, condition )
cds = estimateSizeFactors( cds )
cds = estimateDispersions( cds )
pasillaRes = nbinomTest( cds, "untreated", "treated" )

library(edgeR)
##https://gist.github.com/stephenturner/3745236

datafile = system.file( "extdata/pasilla_gene_counts.tsv", package="pasilla" )
datafile
 
## Read in the data making the row names the first column
counttable <- read.table(datafile, header=T, row.names=1)
head(counttable)
 
## Make metadata data.frame
meta <- data.frame(
  row.names=colnames(counttable),
  condition=c("untreated", "untreated", "untreated", "untreated", "treated", "treated", "treated"),
  libType=c("single", "single", "paired", "paired", "single", "paired", "paired"))
meta$condition <- relevel(meta$condition, ref="untreated")
meta


## Make design matrix
condition <- relevel(factor(meta$condition), ref="untreated")
libType <- factor(meta$libType)
edesign <- model.matrix(~libType+condition)
 
## Make new DGEList, normalize by library size, and estimate dispersion allowing possible trend with average count size
e <- DGEList(counts=counttable)
e <- calcNormFactors(e)
e <- estimateGLMCommonDisp(e, edesign)
e <- estimateGLMTrendedDisp(e, edesign) 
e <- estimateGLMTagwiseDisp(e, edesign)
 
## MDS Plot
plotMDS(e, main="edgeR MDS Plot")
 
## Biological coefficient of variation plot
plotBCV(e, cex=0.4, main="edgeR: Biological coefficient of variation (BCV) vs abundance")
 
## Fit the model, testing the coefficient for the treated vs untreated comparison
efit <- glmFit(e, edesign)
efit <- glmLRT(efit, coef="conditiontreated")
 
## Make a table of results
etable <- topTags(efit, n=nrow(e))$table
etable <- etable[order(etable$FDR), ]
head(etable)
pasillaRes <- etable

@

The \Robject{pasillaRes} object has been saved in the \Rpackage{GeneticsHTSCourse2013} package that accompanies this course.

\usecase
Load the pasilla analysis and familiarise yourself with the contents. Create a data frame of the 100 most significant results.

<<>>=
library(GeneticsHTSCourse2013)
data(pasillaRes)
topHits <- pasillaRes[order(pasillaRes$PValue, decreasing = F)[1:100],]
@

\section{Annotation}
The \Biocpkg{biomaRt} package provides an interface to the biomart website www.biomart.org. The biomaRt package allows the data to be accessed in a consistent manner and without having write complex SQL queries or understand the underlying schema. The databases available through BioMart can be obtained with the \Rfunction{listMarts} function

\subsection{BiomaRt}

\usecase
Get a list of all databases available through the biomaRt package and connect to the Drosophilla one. 
<<>>=
library(biomaRt)
listMarts()

ensembl <- useMart("ensembl", dataset ="dmelanogaster_gene_ensembl")
attr <- listAttributes(ensembl)
head(attr, 10)
@

The function \Rfunction{getBM} is used to build-up queries in biomart. We must specify one or more attributes that we want to retreive. If no values or filters are specified, then all values will be returned.

\usecase
Retrieve the names of all genes for Drosophilla. How many genes are there?

<<>>=
allGenes <- getBM(attributes = c("ensembl_gene_id"), mart = ensembl)
length(allGenes)
@

When restricting the search to particular query items (e.g. genes) we need to specify both filters and values.

\usecase
Retrieve the ensembl, transcript and peptide ID for the first 10 genes

<<>>=
filt <- listFilters(ensembl)
head(filt, 10)
getBM(attributes = attr[1:3, 1], filters = "ensembl_gene_id",
      values = allGenes[1:10], mart = ensembl)

@

For situations when we want to specify multiple filters, we have to supply the values in list form.

\usecase
Retrieve all the genes between 1100000 and 1250000 on chromsome X.
<<>>=
getBM(attributes = "ensembl_gene_id", filters = c("chromosome_name",
"start", "end"), values = list("X", 1100000, 1250000), mart = ensembl)

@

\usecase
Annotate the top hits from the DEseq analysis with their genomic locations and external IDs and create a merged table containing both DESeq results and the annotations.

<<>>=
myInfo <- getBM(attributes = c("flybase_gene_id", "chromosome_name", 
"band", "start_position", "end_position", "external_gene_id"), 
filters = "flybase_gene_id", values = rownames(topHits), mart = ensembl)
myInfo
annotatedHits <- merge(topHits, myInfo, by.x = 0, by.y = 1) 
head(annotatedHits)
@

\subsection{Genome sequences}

\usecase
Load the package that provides the genome representation for Drosophila
<<>>=
library(BSgenome)
available.genomes()
library(BSgenome.Dmelanogaster.UCSC.dm3)
Dmelanogaster
@

\usecase
Retrieve the sequence for chromosome X and subset to get the sequence for gene FBgn0040357.
<<>>=
geneInfo <- getBM(attributes = c("chromosome_name", "start_position", "end_position"),
filters = "ensembl_gene_id", values = "FBgn0040357", ensembl)
gr <- GRanges("chrX", IRanges(geneInfo$start_position, geneInfo$end_position)) 
seq <- getSeq(Dmelanogaster , gr)
seq
@


\usecase
Retrieve sequences for the annotated results from the DEseq analysis and write them out as a Fasta file
<<>>=
myseqs <- getSeq(Dmelanogaster, paste("chr", annotatedHits$chromosome_name, sep = ""), 
annotatedHits$start_position , annotatedHits$end_position)
myseqs
names(myseqs) <- annotatedHits[, 1] 
writeXStringSet(myseqs , file = "topGenes.fa")
@

Various Biostrings operations could now be performed on these sequences. Some suggestions are given below.

<<>>=
substr(myseqs , 1, 10) 
translate(myseqs) 
gcFunction(myseqs)

@


In the analysis of the pasilla dataset, we allocated reads to genes. However, in this section we will explore the ways of counting reads that align to other genomic regions of interest. In this case we will use exons. An example bam file is provided in the \Biocexptpkg{pasillaBamSubset} package, but only reads on chromosome 4 are provided.

\subsection{Organism Annotation}
In order to use the example bam file, which is chromosome 4 only, we will have to get the IDs of genes that occur on chromosome 4. We illustrate this using the \Biocannopkg{org.Dm.eg.db} package.

\usecase
Use the appropriate organism package to retrieve the ensembl IDs of all genes located on chromosome 4 of Drosophilla.

<<>>=
library(org.Dm.eg.db) 
dm <- org.Dm.eg.db
cols(dm)
keytypes(dm)
chr4Genes <- select(dm, cols = "ENSEMBL", keytype = "CHR", keys = "4") 
head(chr4Genes)
chr4ID <- chr4Genes[, 2]
@

\subsection{Transcript packages}

Bioconductor provide a number of pre-built packages that have a database of all transcript information for a particular organism. The package we are going to use is \Biocannopkg{TxDb.Dmelanogaster.UCSC.dm3.ensGene}.

As its name suggests, this package was built from the UCSC tables for the dm3 genome and uses Ensembl genes as an identifier. Convenient functions exist that can return the gene structure as a \Rclass{GRanges} object.

\usecase
Get all the exons for all genes on chromosome 4

<<>>=
library(TxDb.Dmelanogaster.UCSC.dm3.ensGene) 
tx <- TxDb.Dmelanogaster.UCSC.dm3.ensGene
exo <- exonsBy(tx, "gene")
exo
length(exo)
exo4 <- exo[names(exo) %in% chr4ID]
length(exo4)
@

\subsection{Exon Counting}

\usecase
Read the example bam file from the \Biocexptpkg{pasillaBamSubset} package. Verify that only chromosome 4 reads are included in the object

<<>>=
library(pasillaBamSubset)
bam <- readGAlignments(untreated3_chr4(), use.name = TRUE)
table(seqnames(bam))
@
We are now in a position to begin overlapping reads with exons. You should have already seen the \Rfunction{findOverlaps} function that will report the amount of ovelap between two sets of ranges

\usecase
Use the \Rfunction{findOverlaps} function to count the number of reads covering each exon of FBgn0002521.
<<>>=
olaps <- findOverlaps(exo4[["FBgn0002521"]], bam)
olaps
bam[subjectHits(olaps)]
@
\Rfunction{summarizeOverlaps} extends \Rfunction{findOverlaps} by providing options to resolve reads that overlap multiple features. Each read is counted a maximum of once. Different modes of counting are available. See the help page for \Rfunction{summarizeOverlaps} for more details.

\usecase
Repeat the counting for all genes on chromosome 4

<<>>=
olapGenes <- summarizeOverlaps(exo4, bam)
countGenes <- assays(olapGenes)$counts
countGenes
@
The count vector has entries for each gene on chromosome 4. If we wanted counts for all exons, then we would have to supply a slightly different argument to the function with

\usecase
Obtain per-exon counts for each gene
<<>>=
 exonRanges <- unlist(exo4)
olapExon <- summarizeOverlaps(exonRanges , bam)
countExons <- assays(olapExon)$counts
countExons
@
The \Rfunction{summarizeOverlaps} function can also take the location of one or more bam files as an argument. This means that the processing of reads and counting is handled in one step.

\usecase
Obtain counts of chromosome 4 genes in the bam files included with the \Biocexptpkg{pasillaSubset} package.

<<>>=
fls <- c(untreated3_chr4(), untreated1_chr4())
names(fls) <- basename(fls)
multiOlap <- summarizeOverlaps(exo4, BamFileList(fls))
head(assays(multiOlap)$counts)
@

\subsection{Importing Tracks}
Genome browser or other tracks are often distributed as files in the formats gff, bed, wig etc. The package \Biocpkg{rtracklayer} provides infrastrucutre to import files in these formats and create a \Rclass{GRanges} representation.

\usecase
Read the gff file supplied with the pasilla package and use these intervals to produce a table of overlaps.

<<>>=
library(rtracklayer)
gffFile <- "Dmel.BDGP5.25.62.DEXSeq.chr.gff"
gff <- paste(system.file("extdata", package = "pasilla"), gffFile,sep="/")
read.table(gff, nrows = 10, sep = "\t") 
gffRange <- import(gff)
gffRange
gffOverlap <- summarizeOverlaps(gffRange , bam) 
@

\usecase
Find out what tracks are available for the dm3 genome. Download the RepeatMasker track.

<<>>=
mySession <- browserSession()
track.names <- trackNames(ucscTableQuery(mySession))
repeats <- getTable(ucscTableQuery(mySession, track = "rmsk",
table = "rmsk")) 
head(repeats)
@



\section{Visualisation}

\subsection{Exporting tracks}
It is also possible to save the results of a Bioconductor analysis in a browser to enable interactive analysis and integration with other data types, or sharing with collaborators. We shall use the bed format for illustration.

\usecase
Select all significant genes from the pasilla analysis and annotate their genome loca- tion. Create a GRanges representation of the locations.

<<>>=
sigResults <- pasillaRes[which(pasillaRes$PValue< 0.1), ]
myInfo <- getBM(attributes = c("flybase_gene_id", "chromosome_name",
"band", "start_position", "end_position", "external_gene_id"),
filters = "flybase_gene_id", values = rownames(sigResults), mart = ensembl) 
myInfo <- merge(sigResults, myInfo, by.x = 0, by.y = 1)
sigRanges <- GRanges(paste("chr", myInfo$chromosome_name, sep = ""),
IRanges(start = myInfo$start_position , end = myInfo$end_position ,
names = myInfo[, 1]), padj = myInfo$PValue, logfc = myInfo$logFC)
@

Rather than just representing the genomic locations, the .bed format is also able to colour each range according to some property of the analysis (e.g. direction and magnitude of change) to help highlight particular regions of interest. A score can also be displayed when a particular region is clicked-on.

\usecase
Create a score from the p-values and colour scheme based on the fold-change.
<<>>=
Score <- -log10(sigRanges$padj)
rbPal <-colorRampPalette(c("red", "blue"))
logfc <-  pmax(sigRanges$logfc, -3) 
logfc <-  pmin(logfc , 3)
Col <- rbPal(10)[as.numeric(cut(logfc, breaks = 10))]
@

The colours and score can be saved in the \Rclass{GRanges} object and {\tt score} and {\tt itemRgb} columns respectively and will be used to construct the track when the export is called.

\usecase
Export the signifcant results from the DE analysis as a .bed track.

<<>>=
mcols(sigRanges)$score <- Score
mcols(sigRanges)$itemRgb <- Col
export(sigRanges , con = "topHits.bed")
@

The export of other formats such as \textit{gff}, \textit{bedGraph}, \textit{wig} and \textit{bigwig} is also possible with \Biocpkg{rtracklayer}. We can also call the UCSC browser directly to display the tracks. See the \Biocpkg{rtracklayer} vignette for details.


\subsection{ggbio}
We will now take a brief look at one of the visualisation packages in Bioconductor that takes advantages of the GenomicRanges and GenomicFeatures object-types.
The documentation for \Biocpkg{ggbio} is very extensive and contains lots of examples.\newline

{\tt http://www.tengfei.name/ggbio/docs/}

Extra background reading can be found on the ggplot2 website; the package on which the principles of ggbio is based.\newline
{\tt http://ggplot2.org/}

The \Biocpkg{Gviz} package is another Bioconductor package that specialising in genomic visualisations, but we will not explore this package in the course.
A useful function within \Biocpkg{ggbio} is \Rfunction{autoplot}, which will construct an appropriate plot based on the object-type of the input. For example, if we pass a GRanges object to the function it will plot the genomic locations on a linear scale and label chromosome and positional information on the plot. The style of the plot can be changed by the layout argument.

\usecase
Plot the locations of the significant hits from the DE analysis. Try the default, karyogram and circle layouts

<<>>=
library(ggbio)
autoplot(sigRanges)
autoplot(sigRanges[seqnames(sigRanges) == "chrX"])
autoplot(sigRanges , layout = "karyogram")
autoplot(sigRanges , layout = "circle")

@
The Manhattan plot is a common way of visualising genome-wide results, and this is implemented as the \Rfunction{plotGrandLinear} function. We have to supply a value to display on the y-axis. This is done by using the \Rfunction{aes} function, which is inherited from \Rpackage{ggplot2}.

\usecase
Plot the genomic locations of the significant genes on a linear scale. Modify the colour scheme to show whether each gene is up- or down-regulated

<<>>=
plotGrandLinear(sigRanges , aes(y = score)) 
mcols(sigRanges)$Up <- logfc > 0 
plotGrandLinear(sigRanges, aes(y = score, col = Up))
@

\Biocpkg{ggbio} is also able to plot the structure of genes according to a particular model represented by a \Robject{GenomicFeatures} object.

\usecase
Find the gene which has the most exon counts in the pasilla bam. Plot the struture of this gene.

<<>>=
myGene <- names(exonRanges)[which.max(countExons)]
autoplot(tx, which = exo4[[myGene]])
@

\usecase
Plot the number of reads in this region surrounding the gene.

<<>>=
myreg <- flank(reduce(exo4[[myGene]]), 500, both = T)
bamSubset <- bam[bam %over% myreg]
autoplot(bamSubset)
@

\usecase
Repeat the plot, but with a smoothed coverage

<<>>=
autoplot(bamSubset , stat = "coverage")
@

The plots produced by \Biocpkg{ggbio} (and indeed \Rpackage{ggplot2} on which the package is based) are not in standard R graphics format, so techniques like \Rcode{par(mfrow)} for arranging plots on the same page are not applicable. However, \Biocpkg{ggbio} allows the plots to be saved as objects, which can later be arranged by specialist functions such as \Rfunction{tracks}. This function automatically aligns the x-axis of each plot to be on the same scale.

\usecase
Make a combined plot of the sample information and transcripts

<<>>=
p1 <- autoplot(tx, which = myreg)
p2 <- autoplot(bamSubset, stat = "coverage")
tracks(p1, p2)
@

The combined plot can also be embellished by adding an Ideogram to show the location of the region being plotted in relation to the chromosome.

\usecase
Get the ideogram data for the drosophilla genome and add to the plot

<<>>=
plotIdeogram(genome = "dm3")
plotIdeogram(genome = "dm3", subchr = "chr4")
start <- min(start(bamSubset))
end <- max(end(bamSubset))
id <- plotIdeogram(genome = "dm3", subchr = "chr4", zoom.region = c(start,end))
tracks(id, p1, p2)
@

\end{document}